### ISSUE TITLE
server-side-apply doesn't merge but replaces live manifest

### ISSUE THREAD

**Issue by SebastianJ91** [argoproj/argo-cd#17882]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**
When syncing using server-side-apply ArgoCD is not merging desired manifest with live manifest. Instead ArgoCD is replacing the live manifest with the desired manifest.

Before syncing diff view is showing, that the manifests will be merged. But this is not happening.
ArgoCD is breaking the manifest without any warning.

ArgoCD v.2.6.7 was merging the manifests without any problem.

**To Reproduce**

Example based on kiali crd with sources from [https://github.com/SebastianJ91/argocd-server-side-apply/tree/main](https://github.com/SebastianJ91/argocd-server-side-apply/tree/main):

1. `kubectl create namespace kiali-test` to create a testing namespace
2. `kubectl apply -f https://raw.githubusercontent.com/SebastianJ91/argocd-server-side-apply/main/1-manual/1-kiali.io_kialis.yaml` to install crd
3. `kubectl apply -f https://raw.githubusercontent.com/SebastianJ91/argocd-server-side-apply/main/1-manual/2-kiali.io_v1alpha1_kiali.yaml` to install a specific kiali instance manifest
4. `kubectl apply -f https://raw.githubusercontent.com/SebastianJ91/argocd-server-side-apply/main/1-manual/3-argocd-application.yaml` to create a argocd application which is deploying [kiali.yaml](https://github.com/SebastianJ91/argocd-server-side-apply/blob/main/2-argocd-application/kiali.yaml) to patch the just created kiali instance manifest
5. Check ArgoCD application diff
6. Diff shows, that labels are added and _spec.istio_labels.app_label_name_ will be changed to _service.istio-io/canonical-name_

**Expected behavior**

After syncing the live manifest will be merged with the desired manifest and changed to what was shown on diff.
But the live manifest is replaced with the desired manifest.
Diff and the result are differing.

**Screenshots**

diff on 2.11.
![argocd_diff](https://github.com/argoproj/argo-cd/assets/43949592/3be30263-bb9a-4edd-9a48-87d726172ad5)
which is correctly showing only the _spec.istio_labels.app_label_name_ value to change, but nothing to delete

---

live after sync on 2.11.0
![argocd_live_2_11_0](https://github.com/argoproj/argo-cd/assets/43949592/0761a048-1c5a-4256-a95a-023bab922ef1)
which is a copy of desired manifest, which doesn't reflect the diff, as all other existing values were deleted

---

live after sync on 2.6.7
![argocd_live_2_6_7](https://github.com/argoproj/argo-cd/assets/43949592/5466d47a-d57b-4e9b-9524-4f523a0e13ab)
which is a merge of live and desired manifest as shown on diff, as it should be

**Version**

```shell
kubectl exec -n argocd -it argocd-server-79bbfb8ffc-rtx8l -- argocd version
argocd: v2.11.0+ee02027
  BuildDate: 2024-04-18T07:48:17Z
  GitCommit: ee020273c03d2f74fc86e5720994c2a0f92f6a56
  GitTreeState: clean
  GoVersion: go1.21.9
  Compiler: gc
  Platform: linux/amd64
```

**Comment by k-stz** [comment]
We can confirm this also on ArgoCD 2.10.5 (OpenShift GitOps Operator 1.12)
sync-options: ServerSideApply=true does not longer complement the yaml but overwrites it!

**Comment by SebastianJ91** [comment]
After some additional testing I only saw this problem happening if the annotation
`kubectl.kubernetes.io/last-applied-configuration`
is existing before syncing. 
If it isn't because the manifest was created by `kubectl create` instead of `kubectl apply` the manifests are merged fine.

Also testing ServerSideDiff doesn't help.

**Comment by Mangaal** [comment]
I tried to reproduce this issue, and it is still affecting the latest version of ArgoCD. I also attempted to do it manually with kubectl. Here are the steps I followed:

1. kubectl create namespace kiali-test to create a testing namespace.
2. kubectl apply -f https://raw.githubusercontent.com/SebastianJ91/argocd-server-side-apply/main/1-manual/1-kiali.io_kialis.yaml to install the CRD.
3. kubectl apply -f https://raw.githubusercontent.com/SebastianJ91/argocd-server-side-apply/main/1-manual/2-kiali.io_v1alpha1_kiali.yaml to install a specific Kiali instance manifest.
4. kubectl apply --server-side=true -f https://raw.githubusercontent.com/SebastianJ91/argocd-server-side-apply/main/2-argocd-application/kiali.yaml (this is what ArgoCD does when we enable [server-side apply](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-options/#server-side-apply)).

I was testing with Kubernetes (server/client) v1.29.0, and the issue persists. However, when I downgraded Kubernetes (server/client) to v1.22.0, the problem seemed to be resolved.

**Comment by Mangaal** [comment]
ArgoCD under the hood uses kubectl server-side-apply when we enable the sync option [Server-Side Apply](https://argo-cd.readthedocs.io/en/latest/user-guide/sync-options/) and I have received confirmation from the Kubernetes team that the observed behaviour is expected and not a bug. 

Details:

The resource was initially created using client-side apply (CSA). Now, when we attempt to manage and synchronise using ArgoCD with server-side apply (SSA), the live manifest is replaced by the desired manifest as mixing client-side-apply and server-side-apply is not supported as confirmed by the k8s team on the issue ( [#125403](https://github.com/kubernetes/kubernetes/issues/125403) ) that I have created.

Previously, in the older version of ArgoCD, we were using kubernetes and kubectl dependencies that contained a bug (see Kubernetes issue ( [#99003](https://github.com/kubernetes/kubernetes/issues/99003) ). This bug has been fixed in a recent pull request (PR [#112905](https://github.com/kubernetes/kubernetes/pull/112905) ).

After upgrading the dependencies, the behaviour has changed to reflect this fix, resulting in the current operational behaviour.
